assets/js/hooks/cell_editor/live_editor.js

```diff
@@ -566,8 +577,83 @@ class LiveEditor {
       );
     });
   }
+
+  clearDoctestDecorations() {
+    this._doctestDecorations.decorationCollection.clear()
+    this._doctestDecorations.deltaDecorations = []
+  }
+
+  _createDoctestDecoration(line, className) {
+    const lineNumber = Number(line)
+    return {
+      range: new monaco.Range(lineNumber, 1, lineNumber, 1),
+      options: {
+        isWholeLine: true,
+        linesDecorationsClassName: className,
+      }
+    }
+  }
+
+  _appendDoctestDecoration(line, className) {
+    const newDecoration = this._createDoctestDecoration(line, className)
+    this._doctestDecorations.deltaDecorations.push(newDecoration)
+    const decos = this._doctestDecorations.deltaDecorations
+    this._doctestDecorations.decorationCollection.set(decos)
+  }
+
+  _popPushDoctestDecoration(line, className) {
+    const newDecoration = this._createDoctestDecoration(line, className)
+    this._doctestDecorations.deltaDecorations.pop()
+    this._doctestDecorations.deltaDecorations.push(newDecoration)
+    const decos = this._doctestDecorations.deltaDecorations
+    this._doctestDecorations.decorationCollection.set(decos)
+  }
```

Comment:
We can have pop and push separately to avoid duplication. That said, ideally I think we wouldn't rely on the ordering. We can have a single function `putDoctestDecoration` and `.filter` out decorations with the same line, then add the new one, wdyt?

Reply:
sounds good! I will use an object key-ed by the line numbers